internal/internal_workflow.go

```diff
@@ -1356,60 +1355,52 @@ func (h *queryHandler) execute(input []byte) (result []byte, err error) {
 }
 
 // Add increments the WaitGroup counter indicating that
-// a coroutine has been added to the WaitGroup and will be
-// waited on.
+// a coroutine has been added to the WaitGroup.
 //
-// param ctx Context -> workflow context
-//
-// param delta int -> the value to increment the WaitGroup counter by
-//
-// returns WaitGroup
-func (wg *waitGroupImpl) Add(ctx Context, delta int) WaitGroup {
-	if delta < 0 {
-		panic("cannot add negative value %d to WaitGroup")
-	}
-
-	for i := 0; i < delta; i++ {
-		f, _ := NewFuture(ctx)
-		wg.selector = wg.selector.AddFuture(f, func(future Future) {
-			wg.mu.Lock()
-			defer wg.mu.Unlock()
-			wg.n = wg.n - 1
-			if wg.n == 0 {
-				wg.ready = true
-			}
-		}).(*selectorImpl)
-	}
-
-	// increment
-	wg.mu.Lock()
-	defer wg.mu.Unlock()
+// param delta int32 -> the value to increment the WaitGroup counter by
+func (wg *waitGroupImpl) Add(delta int32) {
+	state := wg.waiting
 	wg.n = wg.n + delta
-
-	return wg
+	if wg.n < 0 {
+		panic("negative WaitGroup counter")
+	}
+	if (wg.waiting && delta > 0) && (wg.n == delta) {
+		panic("WaitGroup misuse: Add called concurrently with Wait")
+	}
+	if (wg.n > 0) || (!wg.waiting) {
+		return
+	}
+	if wg.waiting != state {
+		panic("WaitGroup misuse: Add called concurrently with Wait")
+	}
+	if wg.n == 0 {
+		wg.future.Set(false, nil)
+	}
 }
 
 // Done decrements the WaitGroup counter, indicating
 // that a coroutine in the WaitGroup has completed
 func (wg *waitGroupImpl) Done() {
-	wg.mu.Lock()
-	defer wg.mu.Unlock()
-	wg.selector.cases[wg.index].future.Set(true, nil)
-	wg.index = wg.index + 1
+	wg.Add(-1)
 }
 
 // Wait blocks and waits for specified number of couritines to
 // finish executing and then unblocks once the counter has reached 0
 //
 // param ctx Context -> workflow context
 func (wg *waitGroupImpl) Wait(ctx Context) {
-	if wg.n == 0 {
+	if wg.n <= 0 {
 		return
 	}
-	for {
-		wg.selector.Select(ctx)
-		if wg.ready {
-			return
-		}
+	if wg.waiting {
+		panic("WaitGroup is reused before previous Wait has returned")
+	}
+
+	wg.waiting = true
+	if err := wg.future.Get(ctx, &wg.waiting); err != nil {
+		panic(err)
+	}
+	if !wg.waiting {
+		wg.future = &futureImpl{channel: NewChannel(ctx).(*channelImpl)}
```

Comment:
Why not NewFuture?

Reply:
I originally implemented using NewFuture(), but I did not want to deal with the returned Settable as well.  I can switch it back and ignore the Settable if it would be more readable.

Reply:
Future interface doesn't have Set method. So you'll have to store both Future and Settable and use the appropriate interface in Done and Wait methods.

Reply:
Ok.  I will use the interfaces.  What is the reason for using the interfaces vs. the futureImpl struct?

Reply:
I think it is just cleaner to use public APIs.